examples/readline.lua

```diff
@@ -0,0 +1,255 @@
+        drawn_before = false,    -- if the prompt has been drawn
+        fsleep = fsleep,         -- the sleep function to use
+    }, readline_mt)
+
+    return self
+end
+
+function readline:draw(redraw)
+    if redraw or not self.prompt_ready then
+        -- we are at start of prompt
+        self.prompt_ready = true
+        p.column(1)
+        ou.write(tostring(self.prompt))
+    else
+        -- we are at current cursor position, move to start of prompt
+        p.column(self.prompt.olen + 1)
+    end
+    -- write prompt & value
+    local value = tostring(self.value)
+    ou.write(value)
+    -- clear remainder of input size
+    ou.write(string.rep(" ", self.max_length - self.value.olen))
+    -- move to cursor position
+    p.column(self.prompt.olen + self.value.ocursor)
+    ou.flush()
+end
+
+function readline:handle_key(key, keytype)
+    -- registered exit key
+    if exit_keys[key] then
+        return "exit_key"
+    end
+
+    -- registered editting key
+    local handler = self.value[key_names[key]]
+    if handler then
+        handler(self.value)
+        self:draw()
+        return "ok"
+    end
+    -- unregistered ansi sequence
+    if keytype == "ansi" then
+        -- print("unhandled ansi: ", key:sub(2,-1), string.byte(key, 1, -1))
+        bell()
+        return "ok"
+    end
+    -- control character
+    if key < " " then
+        bell()
+        return "ok"
+    end
+    -- maximum length reached
+    if self.value.ilen >= self.max_length then
+        bell()
+        return "ok"
+    end
+    -- add the new received key to the string
+    self.value:add(key)
+    self:draw()
+    return "ok"
+end
+
+--- Processes key input using an event-driven approach.
+-- This function listens for key events and processes them.
+-- If an exit key is pressed, it yields the input value and the exit key.
+-- @event key Triggered when a key is pressed.
+-- @tparam string key The key that was pressed.
+-- @tparam string keytype The type of the key (e.g., "ansi", "control").
+function readline:readkey_co()
+    while true do
+        local key, keytype = t.input.readansi(0.1, self.fsleep) -- blocking sleep
+        if key then
+            local status = self:handle_key(key, keytype)
+            if status == "exit_key" then   -- exit key pressed?
+                coroutine.yield(tostring(self.value), key)
+            elseif status ~= "ok" then
+                error(status)
+            end
+        end
+    end
+end
+
+--- Starts the readline input loop.
+-- This function initializes the input loop for the readline instance.
+-- It uses a coroutine to process key input until an exit key is pressed.
+-- @tparam boolean redraw Whether to redraw the prompt initially.
+-- @treturn string The final input value entered by the user.
+-- @treturn string The exit key that terminated the input loop.
+function readline:run(redraw)
+    self:draw(redraw)
+
+    local co = coroutine.create(function() self:readkey_co() end)
```

Comment:
this coroutine stuff isn't required. If you just drop the `fsleep` function all together, then `readansi` will by default use the sleep function the user set when initializing the terminal.

This means that the user controls behaviour, and there is nothing to do in this lineeditor class.